### DISCUSSION TITLE
Some passed triggers don't work after upgrade from v5.8 to v6.1

### DISCUSSION THREAD

**Discussion by akshaymankar** [concourse/concourse#5856] [Help & Support]
## Summary

After upgrade from v5.8 to v6.1 (via v6.0 and unsuccessful upgrade to v6.3), some passed triggers don't work. Even manually triggering such jobs doesn't make them `get` the latest resource.

## Steps to reproduce

Unfortunately, I haven't gotten around to reproducing this error in a completely new environment as following the steps would be quite cumbersome.

## Expected results

Jobs should continue getting triggered and new resources should be fetched on manual trigger.

## Actual results

Some jobs don't get triggered

## Additional context

The concourse instance was upgraded from v5.8 to v6.0 to make sure everything was fine. An upgrade to concourse v6.3 failed on workers due to some issue with the ansible role we depend on, but the web instance had been upgraded. We rolled back the version to v6.1 and things looked fine, until we noticed this issue.  

My pipeline looks like this:

- Job A gets triggerd on a pull request, puts a semver
- Job B gets the pull request with `passed: [Job A]`, `trigger: true`, `version: every`
- Job B also gets the semver with `passed: [Job A]`, but without `trigger: true` and `version: every`. 
- Job C gets the pull request with `passed: [Job A]`, `trigger: true`, `version: every`
- Job C doesn't get semver at all

For me, Job A and Job B keep getting triggered for each pull request. But Job C doesn't. When I trigger Job C, it gets some old version of the pull request.

I removed the `passed` and `trigger` settings from Job C, set the pipeline, added them back and set the pipeline again. Now, everything works as intended.

I wonder if something has gone wrong with migrations due to the downgrade. It would be nice if there was a guide to downgrade concourse versions.

## Triaging info

* Concourse version: 5.8 -> 6.0 -> 6.3 -> 6.1
* Browser (if applicable): Not applicable
* Did this used to work? 
  The upgrade: never tried before.
  The trigger: yes.

**Comment by jamieklassen** [comment]
Thanks for giving a detailed report of your experience. Until we have a consistent local reproducible, though, I'm going to convert this to a discussion following https://blog.concourse-ci.org/nobody-wants-a-stale-bot/#an-update-on-triage. In the meantime, can you give an actual pipeline YAML or excerpt?

Did you follow https://concourse-ci.org/concourse-web.html#downgrading for the downgrade? Downgrading is not a workflow that's very extensively supported, but we do test some really basic workflows continuously: https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/downgrade:

https://github.com/concourse/ci/blob/342c2822c43c8c52d746d45fe4a1f0f6ee766578/pipelines/concourse.yml#L389-L411

**Reply by akshaymankar** [reply]
> can you give an actual pipeline YAML or excerpt?

Sure, it is here: https://gist.github.com/akshaymankar/e6515f055619b7b28bcd2be652530661
The jobs called `haskell-ormolu-ubuntu` and `haskell-unittests-ubuntu` were not getting triggered.

> Did you follow https://concourse-ci.org/concourse-web.html#downgrading for the downgrade? 

I had not, as my searching skills failed me in finding downgrade instructions. But I have now done this migration, thanks for the link. I will let you know if the I see any issues.

I have more information on the behaviour as I enabled tracing from ATC to jaeger.
I noticed that on every `tick` of the scheduler, the scheduler decided to try and resolve inputs for the job and every time it would vouch for an old build id. The build id was same every time, until at some point it changed inexplicably and then it didn't change until I resolved it. It looked like this in jaeger UI:

![image](https://user-images.githubusercontent.com/649161/86939432-d9d6f600-c141-11ea-91bc-8948208339b8.png)

In the `build_pipes` table, there was no row corresponding to this `from_build_id` and the job. There were rows for other downstream jobs. 

It actually only resolved after I removed the `passed` constraint from the get and let the job run for all the inputs (I also realized that the passed constraint wasn't really necessary in my usecase).
After it was resolved, the events for `concourse/Algorithm.Compute` stopped and they only appeared whenever the `bump` job was successful.

**Reply by clarafu** [reply]
Hm that is really interesting, from first glance I don't see anything complicated within your pipeline config that could result in the Job-c not triggering because the only constraints it really had were one input with passed and version every. I do suspect it might be something to do with the upgrade and downgrade, but I will have to do some more investigation to be sure. Thanks a lot for the super detailed information on this, the more information we have the easier it will be to figure out the problem!

**Reply by akshaymankar** [reply]
Hello, the problem showed up again after the job ran again a couple of times. I think I will just remove the `passed` constraint to get this thing working for us. Before that I will dig a little into it and let y'all know if I find something interesting.

**Reply by akshaymankar** [reply]
I didn't find anything more interesting. I just created the pipeline with a new name, I guess this created new jobs and resource configs. Now everything seems to be ok.

**Reply by akshaymankar** [reply]
So, the issue is back again, even in the new pipeline. 
It seems like it runs a couple of times and then starts failing again. I tried
to look at the code in `group_resolver.go` to come up with some reason as to why
this happens.

I see that [versions.LatestBuildUsingLatestVersion](https://github.com/concourse/concourse/blob/v6.1.0/atc/db/versions_db.go#L390-L421
returns "build 1" of the haskell-ormolu job (the one not getting triggerd), when it has two builds.
If I remove the `limit` clause in the query and also select `rcv.check_order`,
I can see that the `check_order` of "build 2" of the haskell-ormolu job is less than that of "build 1".
None of the jobs here were manually triggered, so I have to conclude that the "bump" got triggered for a higher check order before a lower one, or the check order somehow changed in the meanwhile.

This made me think weather the resource-type was behaving as expected by concourse and I think it doesn't.
It deviates from usual resources in these two ways:
- It returns all versions on first call, when it is suppossed to return only the latest one.
- It completely disregards the `version` field while doing the check and always returns all versions.

I will switch to [telia-oss/github-pr-resource](https://github.com/telia-oss/github-pr-resource) and see if I still have these issues.

**Reply by akshaymankar** [reply]
My pipeline is working fine with the new pr resource-type. I think concourse could do a better job at a badly behaving resource. Please let me know if I should create another issue for this.

**Reply by kirthiprakash** [reply]
Hi @akshaymankar, That's some really nice investigation. I am facing the same problem after upgrading from V5.5.7 to V7.1.0.
I see that you have switched the `git` resource type with `github-pr-resource`. Its been close to a year now and just wanted to check if you are facing any similar issues?

**Reply by akshaymankar** [reply]
@kirthiprakash No, I haven't faced any similar issue since. We are still using version 6.7.4 though.

> I see that you have switched the `git` resource type with `github-pr-resource`.

Also, I never faced the problem with the `git` resource, it was a fork of [jtarchie/github-pullrequest-resource](https://github.com/jtarchie/github-pullrequest-resource).

**Reply by kirthiprakash** [reply]
From my investigation, I found that the auto triggering is failing on downstream jobs because of a bug in the scheduling algorithm. I have filed a bug report for the same here https://github.com/concourse/concourse/issues/6988
